python/pyproject.toml

```diff
@@ -2,15 +2,16 @@
 requires = [
   "setuptools>=55.0",
   "Cython>=3",
-  "pyarrow~=17.0",
+  "numpy>=1.16.6",
```

Comment:
We don't use NumPy explicitly, so it feels wrong that we need it as a dependency. If we remove it, PalletJack will not compile because NumPy is still referenced from Arrow libraries (which seems like a bug in Arrow).

Reply:
This sounded weird to me so I looked into this a bit to understand what's going on. It looks like PyArrow headers have some reference to numpy headers for numpy interop, eg. in https://github.com/apache/arrow/blob/main/python/pyarrow/src/arrow/python/numpy_interop.h, so the numpy headers are required for building against PyArrow.

I don't think that's an Arrow bug, it seems reasonable to require numpy for building against the PyArrow libraries even if numpy isn't a Python runtime dependency.

Reply:
I think you're right, but it's still melting my brain.

Reply:
One thing that's not clear to me is why we compile against the PyArrow headers. I haven't dug much into this but if it's possible to compile against the normal Arrow C++ headers then I don't think this numpy dependency would be needed. But changing that is probably out of scope of this PR.

Reply:
> One thing that's not clear to me is why we compile against the PyArrow headers. I haven't dug much into this but if it's possible to compile against the normal Arrow C++ headers then I don't think this numpy dependency would be needed. But changing that is probably out of scope of this PR.

PyArrow headers are automatically used because of `from pyarrow._parquet cimport *` in the our pyx file.

Reply:
Ah OK, makes sense. I guess we can't avoid needing the numpy headers then.